contracts/meta-staking/src/contract.rs

```diff
@@ -174,32 +171,158 @@ mod execute {
+            None => return Err(ContractError::NoDelegationsForValidator {}),
+        }[0];
 
-        // Withdraw rewards as a submessage
-        let withdraw_msg = SubMsg::reply_on_success(
-            DistributionMsg::WithdrawDelegatorReward { validator },
-            WITHDRAW_REWARDS_REPLY_ID,
-        );
+        if total_accumulated_rewards.amount.is_zero() {
+            return Err(ContractError::ZeroRewardsToSend {});
+        }
+
+        // Get all consumers by a single validator iter
+        let total_consumers_iter = CONSUMERS_BY_VALIDATOR
+            .prefix(&validator)
+            .range(deps.storage, None, None, Order::Ascending)
+            .filter_map(|res| res.ok())
+            .collect::<Vec<(Addr, Uint128)>>();
+
+        // Sum of all delegations so we can count the perc of rewards
+        let total_delegations = total_consumers_iter
+            .iter()
+            .map(|res| {
+                Ok(res.1) // amount = Uint128
+            })
+            .sum::<StdResult<Uint128>>()?;
+
+        // Might be expensive, need to moniter/test
+        total_consumers_iter.iter().for_each(|res| {
+            let (addr, amount) = res;
+
+            // Get the rewards amount of the consumer
+            let perc = (*amount / total_delegations).u128() * (100_u128);
+            let rewards_to_add = (perc * total_accumulated_rewards.amount.u128()) / (100_u128);
+
+            // Should be safe because we loop over consumers
+            let mut consumer = CONSUMERS.load(deps.storage, addr).unwrap();
+
+            consumer.rewards = consumer.rewards.add(Uint128::from(rewards_to_add));
+            consumer.rewards_denom = total_accumulated_rewards.denom.clone();
+
+            CONSUMERS.save(deps.storage, addr, &consumer).unwrap();
+
+            VALIDATORS_REWARDS
+                .update::<_, StdError>(deps.storage, (addr, &validator), |_| Ok(consumer.rewards))
+                .unwrap();
+        });
+
+        // Withdraw rewards from validator
+        let withdraw_msg =
+            CosmosMsg::Distribution(DistributionMsg::WithdrawDelegatorReward { validator });
+
+        Ok(Response::default().add_message(withdraw_msg))
+    }
 
-        // TODO On reply, send funds to consumer contract
+    pub fn withdraw_to_customer(
+        deps: DepsMut,
+        consumer: String,
+    ) -> Result<Response, ContractError> {
+        let consumer_addr = deps.api.addr_validate(&consumer)?;
+
+        if !CONSUMERS.has(deps.storage, &consumer_addr) {
+            return Err(ContractError::NoConsumer {});
+        };
+
+        let mut consumer = CONSUMERS.load(deps.storage, &consumer_addr)?;
 
-        Ok(Response::default().add_submessage(withdraw_msg))
+        if consumer.rewards.is_zero() {
+            return Err(ContractError::ZeroRewardsToSend {});
+        }
+
+        // Get list of rewards by validator, so provider will be able to use it
+        let mut rewards_by_validator: Vec<(String, Uint128)> = vec![];
```

Comment:
Another way we might consider doing this is only withdrawing rewards for single validator. Could have both perhaps?

Reply:
so like add optional validator parameter and only return his reward?

Reply:
> so like add optional validator parameter and only return his reward?

Yeah, may be best to implement as a separate contract method? Similar to how we have `WithdrawRewards` and `WithdrawAllRewards` in the Cosmos SDK?

Reply:
> Yeah, may be best to implement as a separate contract method? Similar to how we have `WithdrawRewards` and `WithdrawAllRewards` in the Cosmos SDK?

Yea I can do that.